src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+
+                isFirstSpinner = false;
+                var state = new State(lockObj);
+                while (true)
+                {
+                    State newState = state;
+                    TryLockResult result = TryLockResult.Spin;
+                    if (newState.HasAnyWaiters)
+                    {
+                        if (newState.ShouldNotPreemptWaiters)
+                        {
+                            return TryLockResult.Wait;
+                        }
+                        if (lockObj.ShouldStopPreemptingWaiters)
+                        {
+                            newState.SetShouldNotPreemptWaiters();
+                            result = TryLockResult.Wait;
+                        }
+                    }
+                    if (result == TryLockResult.Spin)
+                    {
+                        Debug.Assert(!newState.ShouldNotPreemptWaiters);
+                        if (!newState.IsLocked)
+                        {
+                            newState.SetIsLocked();
+                            result = TryLockResult.Locked;
+                        }
+                        else if ((newState.HasAnySpinners && spinCount == 0) || !newState.TryIncrementSpinnerCount())
+                        {
+                            return TryLockResult.Wait;
+                        }
+                    }
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        if (result == TryLockResult.Spin && !state.HasAnySpinners)
+                        {
+                            isFirstSpinner = true;
+                        }
+                        return result;
+                    }
+
+                    state = stateBeforeUpdate;
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static TryLockResult TryLockInsideSpinLoop(Lock lockObj)
+            {
+                // This method is called from inside a spin loop, it must unregister the spinner if the lock is acquired
+
+                var state = new State(lockObj);
+                while (true)
+                {
+                    Debug.Assert(state.HasAnySpinners);
+                    if (!state.ShouldNonWaiterAttemptToAcquireLock)
+                    {
+                        return state.ShouldNotPreemptWaiters ? TryLockResult.Wait : TryLockResult.Spin;
+                    }
+
+                    State newState = state;
+                    newState.SetIsLocked();
+                    newState.DecrementSpinnerCount();
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        return TryLockResult.Locked;
+                    }
+
+                    state = stateBeforeUpdate;
```

Comment:
We have contention here on the state change (not on the lock itself). We are clashing with someone changing the state of the lock. 
It may have negative effects on manycore machines if we keep blasting interlock exchanges in a loop. A small exponential backoff before rereading the state would be useful.

Reply:
An example how a backoff on state change collision can be done: https://github.com/dotnet/runtime/blob/0b263953a25136283d3ac45a174b82e49a562aa1/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Lock.cs#L317

Reply:
There can be a tradeoff in doing backoffs, we'd have to be careful in doing so. Frequent backoffs can relatively deprioritize the thread doing the backoffs, which can lead to the thread being delayed heavily, which may show up as increased latency. We had seen this before in ConcurrentQueue with Sleep(1) where threads doing the Sleep(1) help to reduce contention, but get heavily delayed, and we removed the Sleep(1)s. Eventually, we also ended up removing the much shorter spin-waits in forward-progressing CAS loops since it was yielding better perf. All of the CAS loops in this type are forward-progressing, in the sense of the contending threads, one will always win and make progress, and all other threads will see new state for their next iteration (i.e., they won't fall into a busy-spin situation with no changes to memory seen). It may be possible to introduce occasional backoffs to reduce contention briefly but not always. AwareLock hasn't used spin-waits in the CAS loops for a while now and I haven't seen any issues with that, I'd rather not complicate it until issues are seen.

Reply:
A burst of threads trying to modify the state could get very inefficient to resolve in terms of cache misses - if all but one thread fail and have a cache miss on every iteration. 
With 10 threads trying to fetch and then change state aggressively, we will have 10 cache misses and one thread succeeding, then 9 misses, then 8, ... . It seems like the number of misses can be a s bad as n^2. 

Not having the backoff is a problem in the current lock as well. 
It tries to backoff in most cases of collisions, but misses that on the most common case - when trying and failing to acquire. That part is missing a backoff and results in trying to acquire too greedily. It is one of the reasons the current lock does not scale under heavy contention. (the other reason is that it spins too much)

Reply:
> A burst of threads trying to modify the state could get very inefficient to resolve in terms of cache misses - if all but one thread fail and incur a cache miss on every iteration.
> With 10 threads trying to fetch and then change state aggressively, we will have 10 cache misses and one thread succeeding, then 9 misses, then 8, ... . It seems like the number of misses can be as bad as n^2.

I understand the concern, but it seems very unlikely that this would be a problem in these CAS loops. There are cases where backoffs in CAS loops may be beneficial in some ways, such as in longer CAS loops. Using a backoff in CAS loops comes with tradeoffs. Some of those tradeoffs can be seen with quick tests on throughput and latency on short CAS loops. Using an exponential backoff under frequent contention would make it difficult to predict how long it would take a thread using a longer backoff to complete the CAS loop. I was thinking maybe using a backoff in the CAS loops that register new spinners and waiters could enable a lock-holding thread to release its lock more quickly, but some tests there didn't show an improvement. I think we can wait to see if there's an issue and if it needs fixing in the lock implementation, so far I haven't seen this type of issue in coreclr's AwareLock.